### PR TITLE
Issue/glcore

### PR DESCRIPTION
**Pull request by Tronic**
Change from OpenGL 2.1 compatibility/legacy profile to OpenGL 3.3 core profile.

### CODE REVIEW

**Comment by nieknooijens** [comment]
so basically you rewrote the shaders?

**Comment by Tronic** [comment]
In shaders "in" and "out" replace "attribute" and "varying". texture() is used instead of type-specific texture2D() and texture2DRect() of older GLSL versions, allowing removal of some code (especially since Rect is no longer used at all). Additionally a workaround for a bug in Nvidia drivers (for GPUs that were old even in year 2011) was removed (hopefully Nvidia fixed their drivers by now).

**Comment by Tronic** [comment]
glGetAttribLocation (finding GLSL variables by name) is safer than integer values and performance cannot really be an issue for the (relatively low) number of calls that we make.

And no, I didn't know that anyone was already working on this :O

**Comment by Tronic** [comment]
Regarding performance, the biggest issue with this pull request should be that it updates the VBOs every time they are drawn. I suppose this can affect guitar/drum gameplay when a lot of notes are visible. Didn't do any testing and didn't bother to make it only update when needed.

**Comment by Lord-Kamina** [comment]
I agree re: name vs integers. The point is that the whole point is moot because GLSL allows you to explicitly define attribute locations now, so you don't need getLocation because they'll always be where you said they are. For the most part I get the idea your version is cleaner than mine, but do check mine out and see if you can merge the best of both.

**Comment by Lord-Kamina** [comment]
Here: https://github.com/performous/performous/tree/issue/migrate_to_opengl_33

File: game/glutil.cc [hunk 1]
```diff
@@ -40,13 +52,52 @@ void VertexArray::draw(GLint mode) {
...
+		// Prefix with all currently active GLErrorChecker contexts
+		std::string logmsg = "opengl/error: ";
+		for (auto s: stack) logmsg += s + ": ";
+		logmsg += msg(err) + "\n";
+		std::clog << logmsg << std::flush;
+	}
+	stack.back() = info + " after " + what;
+}
+
+/* static */ std::string GLErrorChecker::msg(GLenum err) {
```
Comment (Baklap4) [inline game/glutil.cc]: Remove the static keyword if it's not needed

File: game/glutil.cc [hunk 2]
```diff
@@ -40,13 +52,52 @@ void VertexArray::draw(GLint mode) {
...
+		case GL_INVALID_OPERATION: return "Invalid operation";
+		case GL_INVALID_FRAMEBUFFER_OPERATION: return "FBO is not complete";
+		case GL_STACK_OVERFLOW: return "Stack overflow";
+		case GL_STACK_UNDERFLOW: return "Stack underflow";
+		case GL_OUT_OF_MEMORY: return "Out of memory";
+		default: return "Unknown error";
+	}
+}
+
+/* static */ thread_local std::vector<std::string> GLErrorChecker::stack;
```
Comment (Baklap4) [inline game/glutil.cc]: Remove the static keyword if it's not needed

Comment (Tronic) [reply game/glutil.cc]: This is a definition for a static member variable. Due to C++ syntax, the static keyword can only be used inside class {} and not again when the variable is defined. The comment is there to make it clear to the reader that this is in fact a static member.

Comment (Baklap4) [reply game/glutil.cc]: alright Granted 👍

**Pull request merged**